devito/function.py

```diff
@@ -266,16 +282,42 @@ def data_allocated(self):
         The allocated data values, that is domain+halo+padding.
 
         Elements are stored in row-major format.
+
+        .. note::
+
+            With this accessor you are claiming that you will modify
+            the values you get back. If you only need to look at the
+            values, use :meth:`data_ro_allocated` instead.
         """
+        self._is_halo_dirty = True
+        self._halo_exchange()
         return self._data
 
     @property
-    def halo(self):
-        return self._halo
+    @_allocate_memory
+    def data_ro_domain(self):
```

Comment:
why do we need read only?

Reply:
dropping this in the following PR. What I actually meant was "access to data without triggering a halo update". So you can safely ignore these methods for now